### PR TITLE
[Don't Merge] WIP

### PR DESCRIPTION
**Pull request by divyaN73**
Changes on bico token for polygon deployement

### CODE REVIEW

File: contracts/bico-token/bico/PolygonBicoToken.sol [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+pragma solidity 0.8.4;
+
+import "./BicoTokenImplementation.sol";
+
+contract PolygonBicoToken is BicoTokenImplementation {
+    string private _revertMsg;
+    bytes32 public constant DEPOSITOR_ROLE = keccak256("DEPOSITOR_ROLE");
+
+    modifier only(bytes32 role) {
```
Comment (livingrockrises) [inline contracts/bico-token/bico/PolygonBicoToken.sol]: This is not required. As we are using AccessControlUpgradeable in BicoImplementation, when you call _setupRole it gets set up and you can just use modifier onlyRole(DEPOSITOR_ROLE)

File: contracts/bico-token/bico/PolygonBicoToken.sol [hunk 2]
```diff
@@ -0,0 +1,57 @@
+// File: contracts/child/ChildToken/UpgradeableChildERC20/UChildERC20.sol
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.4;
+
+import "./BicoTokenImplementation.sol";
+
+contract PolygonBicoToken is BicoTokenImplementation {
```
Comment (livingrockrises) [inline contracts/bico-token/bico/PolygonBicoToken.sol]: This contract should also inherit Initializable imo. holding this thought

File: contracts/bico-token/bico/PolygonBicoToken.sol [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    string private _revertMsg;
```
Comment (livingrockrises) [inline contracts/bico-token/bico/PolygonBicoToken.sol]: _revertMsg is never assigned a value. We can remove this along with modifier where it's used

File: contracts/bico-token/bico/PolygonBicoToken.sol [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            _revertMsg
+        );
+        _;
+    }
+
+    /**
+     * @notice Initialize the contract after it has been proxified
+     * @dev meant to be called once immediately after deployment
+     */
+    function initialize( address beneficiary, address trustedForwarder, address governor, address accessControlAdmin, address pauser, address minter, address childChainManager )
```
Comment (livingrockrises) [inline contracts/bico-token/bico/PolygonBicoToken.sol]: maybe rename this function to polygonBico_init()

File: contracts/bico-token/bico/PolygonBicoToken.sol [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+     * @notice called when token is deposited on root chain
+     * @dev Should be callable only by ChildChainManager
+     * Should handle deposit by minting the required amount for user
+     * Make sure minting is done only by this function
+     * @param user user address for whom deposit is being done
+     * @param depositData abi encoded amount
+     */
+    function deposit(address user, bytes calldata depositData)
+        external
+        // override
```
Comment (livingrockrises) [inline contracts/bico-token/bico/PolygonBicoToken.sol]: yes we don't need override

File: contracts/bico-token/bico/PolygonBicoToken.sol [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        only(DEPOSITOR_ROLE)
+    {
+        uint256 amount = abi.decode(depositData, (uint256));
+        _mint(user, amount);
```
Comment (livingrockrises) [inline contracts/bico-token/bico/PolygonBicoToken.sol:32]: should this necessarily emit Transfer event? their docs says and Rahul said its needed
emit Transfer(address(0), user, amount);

Comment (livingrockrises) [reply contracts/bico-token/bico/PolygonBicoToken.sol:32]: ok _mint() should be emitting this already

File: contracts/bico-token/bico/PolygonBicoToken.sol [hunk 7]
```diff
@@ -0,0 +1,44 @@
+// File: contracts/child/ChildToken/UpgradeableChildERC20/UChildERC20.sol
```
Comment (tomarsachin2271) [inline contracts/bico-token/bico/PolygonBicoToken.sol]: Either give absolute path for this commented line or remove it.

File: contracts/bico-token/bico/PolygonBicoToken.sol [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+contract PolygonBicoToken is Initializable, BicoTokenImplementation {
+    bytes32 public constant DEPOSITOR_ROLE = keccak256("DEPOSITOR_ROLE");
+
+    /**
+     * @notice Initialize the contract after it has been proxified
+     * @dev meant to be called once immediately after deployment
+     */
+    function polygonBico_init( address beneficiary, address trustedForwarder, address governor, address accessControlAdmin, address pauser, address minter, address childChainManager ) public initializer {
+        initialize(beneficiary, trustedForwarder, governor, accessControlAdmin, pauser, minter);
```
Comment (tomarsachin2271) [inline contracts/bico-token/bico/PolygonBicoToken.sol:15]: Need to change in BicoTokenImplementation also to not mint any tokens in initialise function if beneficiary is address(0)

Comment (tomarsachin2271) [reply contracts/bico-token/bico/PolygonBicoToken.sol:15]: Also write test case for this.

**Review by tomarsachin2271 — APPROVED** [review]

**Pull request merged**